op-program/client/interop/consolidate.go

```diff
@@ -33,36 +37,87 @@ func ReceiptsToExecutingMessages(depset depset.ChainIndexFromID, receipts ethtyp
 	return execMsgs, logCount, nil
 }
 
-func RunConsolidation(deps ConsolidateCheckDeps,
-	oracle l2.Oracle,
+func fetchAgreedBlockHashes(oracle l2.Oracle, superRoot *eth.SuperV1) ([]common.Hash, error) {
+	agreedBlockHashes := make([]common.Hash, len(superRoot.Chains))
+	for i, chain := range superRoot.Chains {
+		output := oracle.OutputByRoot(common.Hash(chain.Output), chain.ChainID)
+		outputV0, ok := output.(*eth.OutputV0)
+		if !ok {
+			return nil, fmt.Errorf("unsupported L2 output version: %d", output.Version())
+		}
+		agreedBlockHashes[i] = common.Hash(outputV0.BlockHash)
+	}
+	return agreedBlockHashes, nil
+}
+
+func RunConsolidation(
+	logger log.Logger,
+	bootInfo *boot.BootInfoInterop,
+	l1PreimageOracle l1.Oracle,
+	l2PreimageOracle l2.Oracle,
 	transitionState *types.TransitionState,
 	superRoot *eth.SuperV1,
+	tasks taskExecutor,
 ) (eth.Bytes32, error) {
+	deps, err := newConsolidateCheckDeps(transitionState, superRoot.Chains, l2PreimageOracle)
+	if err != nil {
+		return eth.Bytes32{}, fmt.Errorf("failed to create consolidate check deps: %w", err)
+	}
+
 	var consolidatedChains []eth.ChainIDAndOutput
 
+	agreedBlockHashes, err := fetchAgreedBlockHashes(l2PreimageOracle, superRoot)
+	if err != nil {
+		return eth.Bytes32{}, err
+	}
+
 	for i, chain := range superRoot.Chains {
 		progress := transitionState.PendingProgress[i]
 
-		// TODO(#13776): hint block data execution in case the pending progress is not canonical so we can fetch the correct receipts
-		block, receipts := oracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
+		// It's possible that the optimistic block is not canonical.
+		// So we use the blockDataByHash hint to trigger a block rebuild to ensure that the block data, including receipts, are available.
+		_ = l2PreimageOracle.BlockDataByHash(agreedBlockHashes[i], progress.BlockHash, chain.ChainID)
+
+		optimisticBlock, receipts := l2PreimageOracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
```

Comment:
nit: I wonder if we should combine these into one oracle call? e.g. `ReceiptsByOptimisticBlockHash` Then the call would be more declarative about what we are asking for and the oracle can encapsulate the implementation details of what hints it needs to send.

Reply:
Not a bad idea. Will do this.

Reply:
Logged a [ticket](https://github.com/ethereum-optimism/optimism/issues/14014) to deal with this later.